resolwe_bio/processes/import_data/metadata.py

```diff
@@ -148,6 +173,189 @@ def run(self, inputs, outputs):
         if n_samples < 1:
             self.error("The uploaded table contains no samples.")
 
+        df_data = convert_to_data_frame(str(path))
+
+        sample_columns = {"mS#Sample ID", "mS#Sample name", "mS#Sample slug"}
+        if not sample_columns.intersection(set(df_data.columns)):
+            self.error(
+                f"Missing sample column. The uploaded Orange table should "
+                f"contain one of the following columns: {sorted(sample_columns)}."
+            )
+
+        samples = Entity.filter(collection_id=collections[0].id)
+
+        if "mS#Sample ID" in df_data.columns:
+            col_samples = set([sample.id for sample in samples])
+            sample_header = "mS#Sample ID"
+            intersection = col_samples.intersection(set(df_data[sample_header].values))
+        elif "mS#Sample slug" in df_data.columns:
+            col_samples = set([sample.slug for sample in samples])
+            sample_header = "mS#Sample slug"
+            intersection = col_samples.intersection(set(df_data[sample_header].values))
+        elif "mS#Sample name" in df_data.columns:
+            col_samples = set([sample.name for sample in samples])
+            sample_header = "mS#Sample name"
+            intersection = col_samples.intersection(set(df_data[sample_header].values))
+
+        if not intersection:
+            self.error(
+                "None of the samples listed in the uploaded Sample metadata table "
+                "matches the Samples in the working Collection."
+            )
+
+        dup_samples = df_data[df_data.duplicated(subset=sample_header)][
+            sample_header
+        ].values
+        if set(dup_samples).intersection(col_samples):
+            self.error(
+                f"Duplicated metadata entries {dup_samples} were matched to the Samples "
+                f"contained in the Collection. Please use the metadata upload process that "
+                f"allows for one-to-many relations instead."
+            )
+
+        outputs.table = str(path)
+        outputs.n_samples = n_samples
+        outputs.features = get_features_description(data)
+        outputs.target = get_target_description(data)
+        outputs.n_metas = len(data.domain.metas)
+
+
+class UploadOrangeMetadataOneToMany(Process):
```

Comment:
I am just curious... Could we shorten this class by just subclassing the `UploadOrangeMetadataOneToOne`? And than overwrite the diffs. This would make the diffs stand out more clearly + removed the unnecessary duplication.
Maybe @gregorjerse would know if there are any potential problems with this approach.

Reply:
This would be an ideal solution to avoid duplicated code. Another alternative would also be wrapping the majority of processing code in functions.

Reply:
> Could we shorten this class by just subclassing the UploadOrangeMetadataOneToOne?

Probably the other way around — `UploadOrangeMetadataOneToOne` is stricter than `UploadOrangeMetadataOneToMany` in handling duplicate samples.